src/master/udp_rpc.hpp

```diff
@@ -0,0 +1,231 @@
+/*
+ * Copyright (C) 2008 Search Solution Corporation. All rights reserved by Search Solution.
+ *
+ *   This program is free software; you can redistribute it and/or modify
+ *   it under the terms of the GNU General Public License as published by
+ *   the Free Software Foundation; either version 2 of the License, or
+ *   (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program; if not, write to the Free Software
+ *  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+ *
+ */
+
+/*
+ * udp_rpc.hpp - remote procedure call (RPC) over UDP transport module
+ */
+
+#ifndef _UDP_RPC_HPP_
+#define _UDP_RPC_HPP_
+
+#include "hostname.hpp"
+
+#include <functional>
+#include <map>
+#include <thread>
+
+namespace cubhb
+{
+
+  // type aliases
+  using socket_type = SOCKET;
+  using ipv4_type = std::uint32_t;
+  using port_type = std::uint16_t;
+
+  // constants
+  static const std::size_t BUFFER_SIZE = 4096;
+
+  enum message_type
+  {
+    UNKNOWN_MSG = 0,
+    HEARTBEAT = 1
+  };
+
+  class body_type
+  {
+    public:
+      body_type ();
+      body_type (const char *b, std::size_t size);
+
+      body_type (body_type &&other) noexcept;
+      body_type &operator= (body_type &&other) noexcept;
+
+      // Don't allow copy of body_type
+      body_type (const body_type &other) = delete;
+      body_type &operator= (const body_type &other) = delete;
+
+      bool empty () const;
+      std::size_t size () const ;
+      const char *get_ptr () const;
+
+      template <typename Body>
+      void set (message_type type, const Body &b);
+
+      template <typename Body>
+      void get (message_type &type, Body &body) const;
+
+    private:
+      bool m_use_eblock;
```

Comment:
The versatility of this object type may lead to issues when another developer unfamiliar with the broader design expects that this object may server as an transformer between buffer and object (Body) ; it looks more like an union.
What about splitting it into two classes ?:
- pack_helper (with cubmem::extensible_block, and set and get_ptr methods)
- unpack_helper (with cubmem::block m_block, and get and get_ptr methods)